worksheet_D/oxo.py

```diff
@@ -0,0 +1,176 @@
+BOARD_EDGE_LENGTH = 3
+WINNING_ROW_LENGTH = 3
+
+# Changing the board size obviously requires redefining the --+---+-- drawing section in show().
+# The winning row may not have to be the BOARD_EDGE_LENGTH so this is built in to the checks.
+#
+# oxotest.py does not allow lengths greater than 3 currently. I may tweak it to prove it works.
+#
+# Also would have written as BOARD_HEIGHT and BOARD_WIDTH but the worksheet asks for n * n implying square.
+
+
+class OxoBoard:
+    def __init__(self):
+        """ The initialiser. Initialise any fields you need here. """
+        self.board = {}
+        for x in xrange(BOARD_EDGE_LENGTH):
+            for y in xrange(BOARD_EDGE_LENGTH):
+                self.board.update({(x, y) : 0})
+
+    def get_square(self, x, y):
+        """ Return 0, 1 or 2 depending on the contents of the specified square. """
+        return self.board.get((x, y))
+
+    def set_square(self, x, y, mark):
+        """ If the specified square is currently empty (0), fill it with mark and return True.
+            If the square is not empty, leave it as-is and return False. """
+        if self.get_square(x, y) == 0:
+            self.board[(x, y)] = mark
+            return True
+        else:
+            return False
+
+    def is_board_full(self):
+        """ If there are still empty squares on the board, return False.
+            If there are no empty squares, return True. """
+
+        # Couldn't get this next line to work. Could it work?
+
+        # Missed the return! Cheers Ed!
+
+        return all(value != 0 for value in self.board.values())
```

Comment:
There's an even shorter way! Using the `not in` operator, which tests if a value is not present in a sequence. Try this:
```
return 0 not in self.board.values()
```